src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Internal/ValidationStack.cs

```diff
@@ -0,0 +1,75 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using Microsoft.AspNetCore.Mvc.Internal;
+
+namespace Microsoft.AspNetCore.Mvc.ModelBinding.Internal
+{
+    public class ValidationStack
+    {
+        public int Count => _hashSet?.Count ?? _list.Count;
+
+        // We tested the performance of a list at size 15 and found it still better than hashset, but to avoid a costly
+        // O(n) search at larger n we set the cutoff to 20. If someone finds the point where they intersect feel free to change this number.
+        internal const int CutOff = 20;
+
+        internal List<object> _list { get; } = new List<object>();
+
+        internal HashSet<object> _hashSet { get; set; }
```

Comment:
Rename `_list` and `_hashSet` because they're now properties and all properties should have PascalCase names.